### PR TITLE
Comparison of equivalence relations/partitions in invariants tests

### LINKED ISSUE #415: invariant: cf restricted to code points defines the same partition as scf
From https://github.com/unicode-org/properties/issues/23#issuecomment-1437106706 :

add an invariant test that for all code points C, Case_Folding(C) is either equal to Simple_Case_Folding(C) or has two or more code points

### PR DESCRIPTION
**Pull request by eggrobin**
In a future PR I should redo #439 using this, and this should help with some of the old untested invariants.

Prior to https://github.com/unicode-org/unicodetools/pull/474/commits/bd984ab353ef5de9cd351420da45ed5b6fff2eda, since I did not have #471, the new test failed as follows:
<blockquote>
<p><b># Simple and full case foldings define the same equivalence classes on code points.</b></p>
<p><b># This used not to be true, but was rectified by</b></p>
<p><b># https://www.unicode.org/cgi-bin/GetL2Ref.pl?175-C19.</b></p>
<p>EquivalencesOf $codepoints Case_Folding ⇔ Simple_Case_Folding</p>
<h2><a name='Test_Failure_1'>**** START Test Failure 1 ****</a></h2>
<p>The implication ⇒ is false.</p>
<p>The implication ⇐ is true.</p>
<table class='f'>
<tr><td>
</td><td></td><td>[cf](ΐ) </td><td>=</td><td> [cf](ΐ) </td><td>=</td><td> ΐ </td><td>but</td><td> ΐ </td><td>=</td><td> [scf](ΐ) </td><td>≠</td><td> [scf](ΐ) </td><td>=</td><td> ΐ
</tr></td>
<tr><td>
</td><td></td><td>[cf](ΰ) </td><td>=</td><td> [cf](ΰ) </td><td>=</td><td> ΰ </td><td>but</td><td> ΰ </td><td>=</td><td> [scf](ΰ) </td><td>≠</td><td> [scf](ΰ) </td><td>=</td><td> ΰ
</tr></td>
<tr><td>
</td><td></td><td>[cf](ﬅ) </td><td>=</td><td> [cf](ﬆ) </td><td>=</td><td> st </td><td>but</td><td> ﬅ </td><td>=</td><td> [scf](ﬅ) </td><td>≠</td><td> [scf](ﬆ) </td><td>=</td><td> ﬆ
</tr></td>
</table>
<h2><a name='Test_Failure_1'>**** END Test Failure 1 ****</a></h2>
</blockquote>

This PR therefore fixes #415.

If the examples are uncommented, they produce the following output:

<blockquote>
<p><b>#   Example:</b></p>
<p><b>#       Case-insensitive equivalence of ASCII identifiers can be done indifferently by uppercasing</b></p>
<p><b>#       or lowercasing.</b></p>
<p>           EquivalencesOf \p{Block=Basic Latin} Uppercase_Mapping ⇔ Lowercase_Mapping</p>
<p>There are no counterexamples to ⇔.</p>
<table>
</table>
<p><b>#       This is not true in the broader Latin script (in fact neither implication holds).</b></p>
<p>           EquivalencesOf \p{Script=Latin} Uppercase_Mapping ⇎ Lowercase_Mapping</p>
<p>The implication ⇒ is false.</p>
<p>The implication ⇐ is false.</p>
<table>
<tr><td>
</td><td></td><td>[uc](I) </td><td>=</td><td> [uc](ı) </td><td>=</td><td> I </td><td>but</td><td> i </td><td>=</td><td> [lc](I) </td><td>≠</td><td> [lc](ı) </td><td>=</td><td> ı
</tr></td>
<tr><td>
</td><td></td><td>[uc](S) </td><td>=</td><td> [uc](ſ) </td><td>=</td><td> S </td><td>but</td><td> s </td><td>=</td><td> [lc](S) </td><td>≠</td><td> [lc](ſ) </td><td>=</td><td> ſ
</tr></td>
<tr><td>
</td><td></td><td>[uc](Ṡ) </td><td>=</td><td> [uc](ẛ) </td><td>=</td><td> Ṡ </td><td>but</td><td> ṡ </td><td>=</td><td> [lc](Ṡ) </td><td>≠</td><td> [lc](ẛ) </td><td>=</td><td> ẛ
</tr></td>
<tr><td>
</td><td></td><td>[uc](ﬅ) </td><td>=</td><td> [uc](ﬆ) </td><td>=</td><td> ST </td><td>but</td><td> ﬅ </td><td>=</td><td> [lc](ﬅ) </td><td>≠</td><td> [lc](ﬆ) </td><td>=</td><td> ﬆ
</tr></td>
<tr><td>
K </td><td>=</td><td> [uc](K) </td><td>≠</td><td> [uc](K) </td><td>=</td><td> k </td><td>even though</td><td> k </td><td>=</td><td> [lc](K) </td><td>=</td><td> [lc](K)</td><td></td><td>
</tr></td>
<tr><td>
Å </td><td>=</td><td> [uc](Å) </td><td>≠</td><td> [uc](Å) </td><td>=</td><td> å </td><td>even though</td><td> å </td><td>=</td><td> [lc](Å) </td><td>=</td><td> [lc](Å)</td><td></td><td>
</tr></td>
<tr><td>
SS </td><td>=</td><td> [uc](ß) </td><td>≠</td><td> [uc](ẞ) </td><td>=</td><td> ß </td><td>even though</td><td> ß </td><td>=</td><td> [lc](ß) </td><td>=</td><td> [lc](ẞ)</td><td></td><td>
</tr></td>
</table>
<p><b>#       The simple and full case foldings do not define the same equivalence classes on these</b></p>
<p><b>#       strings:</b></p>
<p>           Let $strings = [ {Straße} {STRASSE} {ᾠδή} {ὨΙΔΉ} {...} ]</p>
<table class='s'>
</table>
<p><span class='bb'>## Total:	5	(omitting 5 from listing)</b></p>
<br>
<p>           EquivalencesOf $strings Case_Folding ⇎ Simple_Case_Folding</p>
<p>The implication ⇒ is false.</p>
<p>The implication ⇐ is true.</p>
<table>
<tr><td>
</td><td></td><td>[cf](STRASSE) </td><td>=</td><td> [cf](Straße) </td><td>=</td><td> strasse </td><td>but</td><td> strasse </td><td>=</td><td> [scf](STRASSE) </td><td>≠</td><td> [scf](Straße) </td><td>=</td><td> straße
</tr></td>
<tr><td>
</td><td></td><td>[cf](ὨΙΔΉ) </td><td>=</td><td> [cf](ᾠδή) </td><td>=</td><td> ὠιδή </td><td>but</td><td> ὠιδή </td><td>=</td><td> [scf](ὨΙΔΉ) </td><td>≠</td><td> [scf](ᾠδή) </td><td>=</td><td> ᾠδή
</tr></td>
</table>
<p><b>#       Specifically, full case folding is coarser than simple case folding.</b></p>
<p>           EquivalencesOf $strings Case_Folding ⇏ Simple_Case_Folding</p>
<p>The implication ⇒ is false.</p>
<table>
<tr><td>
</td><td></td><td>[cf](STRASSE) </td><td>=</td><td> [cf](Straße) </td><td>=</td><td> strasse </td><td>but</td><td> strasse </td><td>=</td><td> [scf](STRASSE) </td><td>≠</td><td> [scf](Straße) </td><td>=</td><td> straße
</tr></td>
<tr><td>
</td><td></td><td>[cf](ὨΙΔΉ) </td><td>=</td><td> [cf](ᾠδή) </td><td>=</td><td> ὠιδή </td><td>but</td><td> ὠιδή </td><td>=</td><td> [scf](ὨΙΔΉ) </td><td>≠</td><td> [scf](ᾠδή) </td><td>=</td><td> ᾠδή
</tr></td>
</table>
<p>           EquivalencesOf $strings Case_Folding ⇐ Simple_Case_Folding</p>
<p>There are no counterexamples to ⇐.</p>
<table>
</table>
</blockquote>

### CODE REVIEW

**Review by macchiati — APPROVED** [review]
Very nice. One thought for the future; feels like the left/right info could be refactored with a new class that handles the info in a common way. But minor.

File: unicodetools/src/main/resources/org/unicode/text/UCD/UnicodeInvariantTest.txt [hunk 1]
```diff
@@ -115,6 +137,8 @@
 Let $foo = \p{ccc=9}
 Let $fii = \p{toNFD=/$foo/}
 
+Let $codepoints = [\u0000-\U00010FFF]
```
Comment (markusicu) [inline unicodetools/src/main/resources/org/unicode/text/UCD/UnicodeInvariantTest.txt:140]:
```suggestion
Let $codepoints = [\u0000-\U0010FFFF]
```

File: unicodetools/src/main/resources/org/unicode/text/UCD/UnicodeInvariantTest.txt [hunk 2]
```diff
@@ -74,6 +74,28 @@
...
+#       This is not true in the broader Latin script (in fact neither implication holds).
+#          EquivalencesOf \p{Script=Latin} Uppercase_Mapping ⇎ Lowercase_Mapping
+#       The simple and full case foldings do not define the same equivalence classes on these
+#       strings:
+#          Let $strings = [ {Straße} {STRASSE} {ᾠδή} {ὨΙΔΉ} {...} ]
+#          EquivalencesOf $strings Case_Folding ⇎ Simple_Case_Folding
+#       Specifically, full case folding is coarser than simple case folding.
+#          EquivalencesOf $strings Case_Folding ⇏ Simple_Case_Folding
+#          EquivalencesOf $strings Case_Folding ⇐ Simple_Case_Folding
+##########################
```
Comment (srl295) [inline unicodetools/src/main/resources/org/unicode/text/UCD/UnicodeInvariantTest.txt:98]: Should these examples appear in the data file?

File: unicodetools/src/main/resources/org/unicode/text/UCD/UnicodeInvariantTest.txt [hunk 3]
```diff
@@ -74,6 +74,28 @@
...
+# EquivalencesOf <unicodeSet> <props> (⇐|⇔|⇒|⇍|⇎|⇏) <props>
+# 
+# Verify that the equivalence of elements of <unicodeSet> up to the left <props> is is implied,
+# equivalent, or implies equivalence up to the right <props>, or verify the negation of these
+# relations.
+#
+# On both sides, <props> is a composition of properties and filters, as in an “In” line.
+#
+#   Example:
+#       Case-insensitive equivalence of ASCII identifiers can be done indifferently by uppercasing
```
Comment (markusicu) [inline unicodetools/src/main/resources/org/unicode/text/UCD/UnicodeInvariantTest.txt:86]: please rephrase “can be done indifferently”...

File: unicodetools/src/main/resources/org/unicode/text/UCD/UnicodeInvariantTest.txt (hunk 2 above)
Comment (eggrobin) [reply unicodetools/src/main/resources/org/unicode/text/UCD/UnicodeInvariantTest.txt:98]: This is not a published data file; the other syntaxes above have examples too.

File: unicodetools/src/main/resources/org/unicode/text/UCD/UnicodeInvariantTest.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (eggrobin) [reply unicodetools/src/main/resources/org/unicode/text/UCD/UnicodeInvariantTest.txt:140]: Noncharacters and even surrogates are codepoints, see D10.

**Review by macchiati — APPROVED** [review]

**Pull request merged**